sql/collations.go

```diff
@@ -46,6 +46,13 @@ var collationStringToID = map[string]CollationID{}
 // CollationID represents the collation's unique identifier. May be safely converted to and from an uint16 for storage.
 type CollationID uint16
 
+// CollationCoercible represents the coercibility of an expression or node. Although the resulting value from the node
+// or expression may be NULL, this interface returns the coercibility as though a NULL would not be returned.
+type CollationCoercible interface {
+	// CollationCoercibility returns the collation and coercibility of the expression or node.
+	CollationCoercibility(ctx *Context) (collation CollationID, coercibility byte, err error)
```

Comment:
OK, I found `ResolveCoercibility` and the link to the mysql docs where this is explained. Makes sense, those docs should go here too.

My issue is that `ResolveCoercibility` is called a total of 10 times. For that, we have added over 100 implementations of `CollationCoercibility`. This seems like a poor trade, and will be a continual tax on new expressions and nodes over time.

I'm open to ideas how to make this a better trade to solve this problem. Basically I want two things:

1) Less boilerplate in nodes / expression (types are fine, those are rare)
2) An easy way for someone adding a new expression or node to get the right behavior. My preference is that this is "do nothing" for most expression types. Embedding a default implementation is worse but still better than what's here.

So think about it, maybe you can come up with a better solution. My thought is that an analyzer rule could potentially cover most of this logic by analyzing every expression tree and then wrapping things in CollatedExpression where needed, which would confine the complexity in one place, rather than spread throughout the entire codebase.